--- make-4.3/src/function.c	2020-01-03 10:11:27.000000000 +0300
+++ make-4.3-patched/src/function.c	2020-01-21 11:31:12.716095500 +0300
@@ -1165,7 +1165,6 @@
 func_sort (char *o, char **argv, const char *funcname UNUSED)
 {
   const char *t;
-  char **words;
   int wordi;
   char *p;
   size_t len;
@@ -1179,22 +1178,23 @@
       ++wordi;
     }
 
-  words = xmalloc ((wordi == 0 ? 1 : wordi) * sizeof (char *));
-
-  /* Now assign pointers to each string in the array.  */
-  t = argv[0];
-  wordi = 0;
-  while ((p = find_next_token (&t, &len)) != 0)
-    {
-      ++t;
-      p[len] = '\0';
-      words[wordi++] = p;
-    }
-
   if (wordi)
     {
+      char **words;
       int i;
 
+      words = xmalloc (wordi * sizeof (char *));
+
+      /* Now assign pointers to each string in the array.  */
+      t = argv[0];
+      wordi = 0;
+      while ((p = find_next_token (&t, &len)) != 0)
+        {
+          ++t;
+          p[len] = '\0';
+          words[wordi++] = p;
+        }
+
       /* Now sort the list of words.  */
       qsort (words, wordi, sizeof (char *), alpha_compare);
 
@@ -1212,10 +1212,9 @@
 
       /* Kill the last space.  */
       --o;
+      free (words);
     }
 
-  free (words);
-
   return o;
 }
 
