--- a/src/main.c1	2022-03-24 10:26:41.428383000 +0300
+++ b/src/main.c	2022-03-24 10:26:56.381613200 +0300
@@ -279,6 +279,8 @@
 
 int warn_undefined_variables_flag;
 
+static char *warn_undefined_variables_option = 0;
+
 /* If nonzero, always build all targets, regardless of whether
    they appear out of date or not.  */
 
@@ -387,7 +389,8 @@
   -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE\n\
                               Consider FILE to be infinitely new.\n"),
     N_("\
-  --warn-undefined-variables  Warn when an undefined variable is referenced.\n"),
+  --warn-undefined-variables[=TYPE]\n\
+                              Warn when an undefined variable is referenced.\n"),
     NULL
   };
 
@@ -477,7 +480,7 @@
     { CHAR_MAX+3, flag, &trace_flag, 1, 1, 0, 0, 0, "trace" },
     { CHAR_MAX+4, flag_off, &print_directory_flag, 1, 1, 0, 0,
       &default_print_directory_flag, "no-print-directory" },
-    { CHAR_MAX+5, flag, &warn_undefined_variables_flag, 1, 1, 0, 0, 0,
+    { CHAR_MAX+5, string, &warn_undefined_variables_option, 1, 1, 0, "noargs", 0,
       "warn-undefined-variables" },
     { CHAR_MAX+7, string, &sync_mutex, 1, 1, 0, 0, 0, "sync-mutex" },
     { CHAR_MAX+8, flag_off, &silent_flag, 1, 1, 0, 0, &default_silent_flag,
@@ -825,6 +828,24 @@
 #endif
 }
 
+static void
+decode_warn_undefined_flags (void)
+{
+  if (warn_undefined_variables_option)
+    {
+      if (streq (warn_undefined_variables_option, "all"))
+        warn_undefined_variables_flag = WARN_UNDEFINED_BASE | WARN_UNDEFINED_ARGS;
+      else if (streq (warn_undefined_variables_option, "noargs"))
+        warn_undefined_variables_flag = WARN_UNDEFINED_BASE;
+      else if (streq (warn_undefined_variables_option, "args"))
+        warn_undefined_variables_flag = WARN_UNDEFINED_ARGS;
+      else
+        OS (fatal, NILF,
+            _("unknown warn-undefined-variables type '%s'"),
+            warn_undefined_variables_option);
+    }
+}
+
 #ifdef WINDOWS32
 
 #ifndef NO_OUTPUT_SYNC
@@ -3242,6 +3263,7 @@
   /* If there are any options that need to be decoded do it now.  */
   decode_debug_flags ();
   decode_output_sync_flags ();
+  decode_warn_undefined_flags ();
 
   /* Perform any special switch handling.  */
   run_silent = silent_flag;
--- a/src/variable.c1	2022-03-24 10:26:45.350318800 +0300
+++ b/src/variable.c	2022-03-24 10:26:56.397144800 +0300
@@ -1920,3 +1920,22 @@
   free (path);
 }
 #endif
+
+void
+show_undefined_variable_warn (const char *name, size_t length)
+{
+  if (!(warn_undefined_variables_flag & WARN_UNDEFINED_BASE) ||
+      !(warn_undefined_variables_flag & WARN_UNDEFINED_ARGS))
+    {
+      /* $(call ...) arguments are named as $1,$2,$3 and so on,
+         check if variable name is numeric.  */
+      size_t i;
+      for (i = 0; i < length && ISDIGIT(name[i]); ++i);
+
+      if ((i == length && !(warn_undefined_variables_flag & WARN_UNDEFINED_ARGS)) ||
+          (i != length && !(warn_undefined_variables_flag & WARN_UNDEFINED_BASE)))
+          return;
+    }
+  error (reading_file, length, _("warning: undefined variable '%.*s'"),
+         (int)(length), name);
+}
--- a/src/variable.h1	2022-03-24 10:26:48.975456800 +0300
+++ b/src/variable.h	2022-03-24 10:26:56.397144800 +0300
@@ -238,13 +238,16 @@
 #define undefine_variable_global(n,l,o) \
           undefine_variable_in_set((n),(l),(o),NULL)
 
+void show_undefined_variable_warn (const char *name, size_t length);
+
+#define WARN_UNDEFINED_BASE  (0x001)
+#define WARN_UNDEFINED_ARGS  (0x002)
+
 /* Warn that NAME is an undefined variable.  */
 
 #define warn_undefined(n,l) do{\
                               if (warn_undefined_variables_flag)        \
-                                error (reading_file, (l),               \
-                                       _("warning: undefined variable '%.*s'"), \
-                                       (int)(l), (n));                  \
+                                show_undefined_variable_warn (n, l);    \
                               }while(0)
 
 char **target_environment (struct file *file);
