--- make-4.3/src/function.c	2020-01-03 10:11:27.000000000 +0300
+++ make-4.3-patched/src/function.c	2020-01-21 11:31:12.716095500 +0300
@@ -1047,7 +1047,7 @@
               wp = hash_find_item (&a_word_table, &a_word_key);
               while (wp)
                 {
-                  wp->matched |= 1;
+                  wp->matched = 1;
                   wp = wp->chain;
                 }
             }
@@ -1061,7 +1061,7 @@
       for (wp = wordhead; wp != 0; wp = wp->next)
         if (is_filter ? wp->matched : !wp->matched)
           {
-            o = variable_buffer_output (o, wp->str, strlen (wp->str));
+            o = variable_buffer_output (o, wp->str, wp->length);
             o = variable_buffer_output (o, " ", 1);
             doneany = 1;
           }
