--- make-4.4.1-orig/src/main.c	2023-02-21 08:03:51.000000000 +0300
+++ make-4.4.1/src/main.c	2024-09-18 10:15:40.935608300 +0300
@@ -183,8 +183,8 @@
 
 /* Nonzero means do not use any of the builtin rules (-r) / variables (-R).  */
 
-int no_builtin_rules_flag = 0;
-int no_builtin_variables_flag = 0;
+int no_builtin_rules_flag = 1;
+int no_builtin_variables_flag = 1;
 
 /* Nonzero means keep going even if remaking some file fails (-k).  */
 
@@ -284,7 +284,7 @@
 /* If nonzero, we should print a warning message
    for each reference to an undefined variable.  */
 
-int warn_undefined_variables_flag;
+int warn_undefined_variables_flag = 1;
 
 static char *warn_undefined_variables_option = 0;
 
