--- a/src/job.c1	2022-03-24 10:53:43.756782000 +0300
+++ b/src/job.c	2022-03-24 10:57:56.147239900 +0300
@@ -2079,7 +2079,8 @@
   /* sub_proc.c is limited in the number of objects it can wait for. */
   if (process_table_full ())
     return 1;
-#endif
+  return 0;
+#else /* !WINDOWS32 */
 
   if (max_load_average < 0)
     return 0;
@@ -2187,6 +2188,7 @@
                 guess, load, max_load_average));
 
   return guess >= max_load_average;
+#endif /* !WINDOWS32 */
 #endif
 }
 
diff --git a/build_w32.bat b/build_w32.bat
index 653841d7..b1792c49 100755
--- a/build_w32.bat
+++ b/build_w32.bat
@@ -270,7 +270,6 @@ call :Compile src/w32/subproc/sub_proc
 call :Compile src/w32/subproc/w32err
 call :Compile lib/fnmatch
 call :Compile lib/glob
-call :Compile lib/getloadavg
 
 :: Compile dirent unless it is supported by compiler library (like with gcc).
 if "%DIRENT%" == "Y" call :Compile src\w32\compat\dirent
