--- make-4.3/src/w32/subproc/sub_proc.c	2020-01-03 10:11:27.000000000 +0300
+++ make-4.3-patched/src/w32/subproc/sub_proc.c	2020-01-21 11:23:12.254614700 +0300
@@ -645,6 +645,11 @@
                                 &bytes_returned, 0) == FALSE || bytes_returned < 2) {
 
                         pproc->last_err = GetLastError();
+                        if (!pproc->last_err) {
+                                fprintf (stderr, "ReadFile failed to read more than 2 bytes of input (read %d bytes from %s).\n",
+                                         bytes_returned, exec_fname);
+                                pproc->last_err = ERROR_BAD_FORMAT;
+                        }
                         pproc->lerrno = E_IO;
                         CloseHandle(exec_handle);
                         return(-1);
