--- make-4.3/src/function.c	2020-01-03 10:11:27.000000000 +0300
+++ make-4.3-patched/src/function.c	2020-01-21 12:17:35.513262400 +0300
@@ -1117,23 +1125,26 @@
   char *msg, *p;
   size_t len;
 
-  /* The arguments will be broken on commas.  Rather than create yet
-     another special case where function arguments aren't broken up,
-     just create a format string that puts them back together.  */
-  for (len=0, argvp=argv; *argvp != 0; ++argvp)
-    len += strlen (*argvp) + 2;
-
-  p = msg = alloca (len + 1);
-  msg[0] = '\0';
-
-  for (argvp=argv; argvp[1] != 0; ++argvp)
+  msg = argv[0];
+  if (argv[1] != 0)
     {
+      /* The arguments will be broken on commas.  Rather than create yet
+         another special case where function arguments aren't broken up,
+         just create a format string that puts them back together.  */
+      for (len=0, argvp=argv; *argvp != 0; ++argvp)
+        len += strlen (*argvp) + 2;
+
+      p = msg = alloca (len + 1);
+
+      for (argvp=argv; argvp[1] != 0; ++argvp)
+        {
+          strcpy (p, *argvp);
+          p += strlen (*argvp);
+          *(p++) = ',';
+          *(p++) = ' ';
+        }
       strcpy (p, *argvp);
-      p += strlen (*argvp);
-      *(p++) = ',';
-      *(p++) = ' ';
     }
-  strcpy (p, *argvp);
 
   switch (*funcname)
     {
