diff --git a/src/function.c b/src/function.c
index 38ab9667..a7b8bd26 100644
--- a/src/function.c
+++ b/src/function.c
@@ -1207,7 +1207,6 @@ static char *
 func_sort (char *o, char **argv, const char *funcname UNUSED)
 {
   const char *t;
-  char **words;
   int wordi;
   char *p;
   size_t len;
@@ -1221,22 +1220,23 @@ func_sort (char *o, char **argv, const char *funcname UNUSED)
       ++wordi;
     }
 
-  words = xmalloc ((wordi == 0 ? 1 : wordi) * sizeof (char *));
-
-  /* Now assign pointers to each string in the array.  */
-  t = argv[0];
-  wordi = 0;
-  while ((p = find_next_token (&t, &len)) != 0)
-    {
-      ++t;
-      p[len] = '\0';
-      words[wordi++] = p;
-    }
-
   if (wordi)
     {
+      char **words;
       int i;
 
+      words = xmalloc (wordi * sizeof (char *));
+
+      /* Now assign pointers to each string in the array.  */
+      t = argv[0];
+      wordi = 0;
+      while ((p = find_next_token (&t, &len)) != 0)
+        {
+          ++t;
+          p[len] = '\0';
+          words[wordi++] = p;
+        }
+
       /* Now sort the list of words.  */
       qsort (words, wordi, sizeof (char *), alpha_compare);
 
@@ -1254,10 +1254,9 @@ func_sort (char *o, char **argv, const char *funcname UNUSED)
 
       /* Kill the last space.  */
       --o;
+      free (words);
     }
 
-  free (words);
-
   return o;
 }
 
